pkg/netlink/netlink_watcher.go

```diff
@@ -0,0 +1,2059 @@
+		if err != nil {
+			log.Println("error-", err)
+			return routeList{}
+		}
+		RouteData = append(RouteData, ri)
+	}
+	route := ParseRoute(v, RouteData, t)
+	return route
+}
+
+// readRouteFromIP reads the routes from ip
+func readRouteFromIP(v *infradb.Vrf) {
+	var Rl routeList
+	var rm []routeCmdInfo
+	var Rt1 int
+	for _, routeSt := range v.Metadata.RoutingTable {
+		Rt1 = int(*routeSt)
+		Raw, err := nlink.ReadRoute(ctx, strconv.Itoa(Rt1))
+		if err != nil {
+			log.Printf("netlink: Err Command route\n")
+			return
+		}
+		Rl = cmdProcessRt(v, Raw, Rt1)
+		for _, r := range Rl.RS {
+			addRoute(r)
+		}
+	}
+	nl := getNeighborRoutes() // Add extra routes for Resolved neighbors on connected subnets
+	for i := 0; i < len(nl); i++ {
+		rm = append(rm, nl[i])
+	}
+	nr := ParseRoute(v, rm, 0)
+	for _, r := range nr.RS {
+		addRoute(r)
+	}
+}
+
+// readRoutes reads the routes
+func readRoutes(v *infradb.Vrf) {
+	readRouteFromIP(v)
+}
+
+func notifyAddDel(r interface{}, event string) {
+	log.Printf("netlink: Notify event: %s\n", event)
+	EventBus.Publish(event, r)
+}
+
+func (nexthop *NexthopStruct) deepEqual(nhOld *NexthopStruct, nc bool) bool {
+	switch {
+	case nexthop.Vrf.Name != nhOld.Vrf.Name:
+		return false
+	case nexthop.Weight != nhOld.Weight:
+		return false
+	case nexthop.ID != nhOld.ID:
+		return false
+	case nexthop.Key != nhOld.Key:
+		return false
+	case nexthop.Local != nhOld.Local:
+		return false
+	case !reflect.DeepEqual(nexthop.Metadata, nhOld.Metadata):
+		return false
+	case nexthop.Metric != nhOld.Metric:
+		return false
+	case nexthop.Scope != nhOld.Scope:
+		return false
+	case nexthop.Resolved != nhOld.Resolved:
+		return false
+	case nexthop.Protocol != nhOld.Protocol:
+		return false
+	case nexthop.NhType != nhOld.NhType:
+		return false
+	case !reflect.DeepEqual(nexthop.nexthop, nhOld.nexthop):
+		return false
+	}
+	if nc {
+		if len(nexthop.RouteRefs) != len(nhOld.RouteRefs) {
+			return false
+		}
+		for i := range nexthop.RouteRefs {
+			ret := nexthop.RouteRefs[i].deepEqual(nhOld.RouteRefs[i], false)
+			if !ret {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (route *RouteStruct) deepEqual(routeOld *RouteStruct, nc bool) bool {
+	switch {
+	case route.Vrf.Name != routeOld.Vrf.Name:
+		return false
+	case !reflect.DeepEqual(route.Route0, routeOld.Route0):
+		return false
+	case route.Key != routeOld.Key:
+		return false
+	case route.NlType != routeOld.NlType:
+		return false
+	case !reflect.DeepEqual(route.Metadata, routeOld.Metadata):
+		return false
+	}
+	if nc {
+		if len(route.Nexthops) != len(routeOld.Nexthops) {
+			return false
+		}
+		return route.Nexthops[0].deepEqual(routeOld.Nexthops[0], false)
+	}
+	return true
+}
+
+func deepCheck(v1 interface{}, v2 interface{}) bool {
+	switch t := v1.(type) {
+	case *RouteStruct:
+		return t.deepEqual(v2.(*RouteStruct), true)
+	case *NexthopStruct:
+		return t.deepEqual(v2.(*NexthopStruct), true)
+	}
+	return true
```

Comment:
so if we have v1 and v2 of different types,t hen we return true
looks suspicious

Reply:
fixed

Reply:
but what if v1 and v2 of the same type but not `RouteStruct` or `NexthopStruct`. It is still true

Reply:
Done, we now check for 4 possible structs